File: HLC01_03Java/src/main/java/org/example/juegoRol03/Personaje.java
package org.example.juegoRol03;

public abstract class Personaje {
    public abstract String getName();
    public abstract int getHealth();
    public abstract void setHealth(int health);
    public abstract int getDamage();

    public void receiveDamage(int damage) {
        System.out.println(getName() + " recibe " + damage + " de daño");
        setHealth(getHealth() - damage);
        if (!isAlive()) {
            die();
        }
    }

    public void attack(Personaje target) {
        System.out.println(getName() + " ataca con " + getDamage() + " de daño");
        target.receiveDamage(getDamage());
    }

    public boolean isAlive() {
        return getHealth() > 0;
    }

    private void die() {
        System.out.println(getName() + " ha muerto!");
    }

}

File: HLC01_03Kotlin/src/main/kotlin/juegoRol03/Personaje.kt
package org.example.juegoRol03

abstract class Personaje {
    abstract val name: String
    abstract var health: Int
    abstract val damage: Int


    fun receiveDamage(damage: Int) {
        println("$name recibe $damage de daño")
        this.health -= damage
        if (!isAlive()) {
            die()
        }
    }


    fun attack(target: Personaje) {
        println("$name ataca con $damage de daño")
        target.receiveDamage(this.damage)
    }


    fun isAlive(): Boolean {
        return this.health > 0
    }


    private fun die() {
        println("$name ha muerto!")
    }

}